---
title: "Demonstrating `clusterProfiler` for term enrichment"
author: "Darach"
date: "`r Sys.Date()`"
---

After measuring some value for many genes (perhaps genome-wide), 
you may find yourself wondering "what biological knowledge is 
associated with the values I am measuring?" One strategy would be
to look at each gene from the top, and see what knowledge (or perhaps
GO terms) are associated with that gene. 
Another would be to define a top
of the list set and see what is in common with those genes, compared
to the others (ex: GO term enrichment test). Another is to use a GSEA
analysis[^gsea] to see what terms are associated with one end of the list
or the other.

[^gsea]: Details of GSEA 
[here](https://bioconductor.org/packages/release/bioc/vignettes/DOSE/inst/doc/GSEA.html),
described by the author of the implementation we use. Very informative!

Here, we'll demonstrate how to do this in R using the nice package
[`clusterProfiler`](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)
that implements these analyses at multiple levels of
complexity. We'll start with a toy data-set, then work our way up.

A vingette of the package is [here](https://www.bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html),
and is already quite good. The purpose of this document is to 
demonstrate its use more simply with regards to
the toy dataset to explore the fundamental functions, and to
specifically demonstrate its use in yeast. If you're curious, you
might find the simulations to be useful for developing and testing
your understanding of the process. We don't use the
custom visualizations provided, and instead brew our own.
This document also builds off the nice but short
[blog post](http://guangchuangyu.github.io/2015/05/use-clusterprofiler-as-an-universal-enrichment-analysis-tool/)
demonstrating arbitrary mapping, but I wanted to use toy datasets to
play with it a bit more.

Also, there's just a lot of good stuff about GO terms and 
bioinformatics tools on his blog, so things like
[https://guangchuangyu.github.io/2014/08/why-clusterprofiler-fails/](https://guangchuangyu.github.io/2014/08/why-clusterprofiler-fails/) : 
"Software that never 'fail' is untrustable."

# First aquaintances with the analysis, via simulations

To use this, you're going to need to install the libraries
`clusterProfiler` from Bioconductor. You'll need to run 
`source("https://bioconductor.org/biocLite.R")` to get the installer
for the Bioconductor packages. 

I am also going to make use of `tidyverse` because it's handy, 
and I'm assuming you're familiar with it.

```{r,libs,cache=F,warning=F,message=F}
library(tidyverse)
```

I'm going to make a fake dataset, with 26 terms (letters of the 
english alphabet) each associated with 200 genes each.
Each gene has a measurement that is drawn from a normal distribution. 
Terms `A`, `B`, and `C` will have an increased mean of adding `bias`. 
First, I'll make a function that gives us some fake data:

```{r, make-fake-data, cache=T}

makeFakeData <- function(bias=0,bias_sd=0,
                         terms=toupper(letters),genes_per_term=200,
                         terms_to_bias=c("A","B","C")) {
  all_data <- data.frame(
    Term=sort(rep(terms,genes_per_term))
    )
  all_data$GeneID <- 1:nrow(all_data)
  set.seed(1)
  all_data$Measurement <- sapply(all_data$Term,
    function(x){
      if (x %in% terms_to_bias) { 
        return( rnorm(n=1,mean=0+bias,sd=1+bias_sd) )
      } else { 
        return( rnorm(n=1,mean=0,sd=1) )
      }
    })
  return(as_tibble(all_data))
}
head(makeFakeData(bias=1))

g <- function(x) {
  ggplot(x)+aes(x=Term,y=Measurement)+geom_boxplot(outlier.size=0)+
    geom_dotplot(binaxis="y",stackdir="center",binwidth=0.05)
}
g(makeFakeData(bias=0))
g(makeFakeData(bias=0.5))
```

So then we look for enrichment of the term in sets. I'm going to make
fake data with bias 0.5, and then ask what terms are 
enriched in the top 500 genes. To do this, I'm going to split
the fake data set into `datar` of the named and sorted vector of 
measurements, and then make a `term_gene_mapping` for 
`clusterProfiler` to make use of:

```{r, split-fake-data, cache=T}
fake_datar_full <- makeFakeData(bias=0.5,terms_to_bias=c("A","B","C"))

fake_datar <- setNames(object=fake_datar_full$Measurement,
    nm=fake_datar_full$GeneID) 

term_gene_mapping <- fake_datar_full %>% select(.,Term,GeneID) %>%
  mutate_all(.,.funs=as.character) 

head(fake_datar)
head(term_gene_mapping)
```

Now with fake data,
we're going to use the `enricher()` function here to
see how enriched these terms are. 
I've specified as many options
as possible, because that's good practice for these kinds of things.
Let's talk about the arguments:

`gene`
  ~ This is the un-ranked set of genes to take a look at. Below, I
    just use an arbitary cutoff. Importantly, these are the `names()`
    of the genes, not their measurements.

`pAdjustMethod`
  ~ What method to use? See the manpage `?clusterProfiler::enricher` 
    for choices

`pvalueCutoff`
  ~ This is a cutoff applied to the adjusted pvalue.

`qvalueCutoff`
  ~ This is a cutoff applied to the calculated qvalue. What's the
    difference from the adjusted p-value? Frankly, I'm a bit confused
    myself, as I thought a q-value was calculating by adjusting 
    p-value to account for multiple testing such that you've then
    got a FDR which I thought was synonymous with a q-value. 

#Look it up in the paper paywalled

`universe`
  ~ This is the background set, so I just put in all the terms.

`minGSSize`
  ~ This is the minimum considered gene set size...

`maxGSSize` 
  ~ ... and this is of course the maximum. These two parameters should
    be tuned based on your biological question, are you interested
    on specific complexes of ~3 genes, or are you interested in
    very broad trends like "translation". 

`TERM2GENE`
  ~ This is the mapping data.frame (or tibble apparently) from above,
    the order of columns is important, not the names.

`TERM2NAME`
  ~ This is for pretty printing, and I don't use it.

```{r, top-slice, cache=T}
enricher_result <- clusterProfiler::enricher(
  gene=names(sort(fake_datar,decreasing=T)[0:500]),
  pvalueCutoff=0.05,
  qvalueCutoff=0.05,
  pAdjustMethod="fdr",
  universe=term_gene_mapping$GeneID,
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
enricher_result 
```

You can coerce the result to a data.frame or tibble to get a table
of values back:

```{r,tibbled,cache=T}
as.tibble(enricher_result)
```

And here's what the object really looks like under `str`.

```{r,strd,cache=T}
str(enricher_result)
```

We see that there's a result of our enriched terms, and that if
we coerce the result to a tibble (or data.frame), we get a nice
result table back. 

What if it's a much more slight bias? Note I just re-generate the
`datar` vector.

```{r, slight-effect-enricher, cache=T}
fake_datar_full <- makeFakeData(bias=0.3,terms_to_bias=c("A","B","C"))
fake_datar <- setNames(object=fake_datar_full$Measurement,
    nm=fake_datar_full$GeneID)

enricher_result <- clusterProfiler::enricher(
  gene=names(sort(fake_datar,decreasing=T)[1:500]),
  pvalueCutoff=0.05,
  qvalueCutoff=0.05,
  pAdjustMethod="fdr",
  universe=term_gene_mapping$GeneID,
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
as.tibble(enricher_result)
```

Subtle effects can be pulled out by gene-set enrichment analyses,
for better (faint biological signal) or worse (technical bias).

`geneList`
  ~ This is the "ordered ranked geneList". Make sure it is so, the
    function will check for you.

`nPerm`
  ~ Number of permutations to use for the bootstrapping. You'll need
    to crank this up to ~1e7 for actual running it, but you can use
    smaller numbers to test implementation (with large pvalue 
    cutoffs).

`minGSSize`
  ~ This is the minimum considered gene set size...

`maxGSSize`
  ~ ... and this is of course the maximum. These two parameters should
    be tuned based on your biological question, are you interested
    on specific complexes of ~3 genes, or are you interested in
    very broad trends like "translation". 

`pAdjustMethod`
  ~ What method to use? See the manpage `?clusterProfiler::enricher` 
    for choices

`pvalueCutoff`
  ~ This is a cutoff applied to the adjusted pvalue.

`exponent`
  ~ This has to do with the weight of each step. Consult the 
    implementation for more info. I don't mess with it.

`TERM2GENE`
  ~ This is the mapping data.frame (or tibble apparently) from above,
    the order of columns is important, not the names.

`TERM2NAME`
  ~ This is for pretty printing, and I don't use it.

```{r, slight-effect-gsea, cache=T}
GSEA_result <- clusterProfiler::GSEA(
  geneList=sort(fake_datar,decreasing=T),
  nPerm=1e6,
  pvalueCutoff=0.05,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
as.tibble(GSEA_result) 
```

It takes longer, but you don't have to pick a cut-off. 
What about if it's biased towards the other end of the set?
Below, I crank up the bias so I can run it with lower 
permutations[^perm].

[^perm]: Your sensitivity depends on permutations! If you're looking
for p-values of 0.05 with 1000 terms, then you'd need to have at
least enough sampling to estimate that something is 0.05/1000 rare.
That means you need at least 20,000 permutations. More is better, but
of course slower.

```{r, slight-neg-effect-gsea, cache=T}
fake_datar_full <- makeFakeData(bias=-2,terms_to_bias=c("A","B","C"))
fake_datar <- setNames(object=fake_datar_full$Measurement,
    nm=fake_datar_full$GeneID)

GSEA_result <- clusterProfiler::GSEA(
  geneList=sort(fake_datar,decreasing=T),
  nPerm=1e5,
  pvalueCutoff=0.05,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
as.tibble(GSEA_result) 
```

So a negative enrichment score. I believe this is for terms that are
de-enriched from one side. To test this, I've got a `bias_sd`
parameter in the simulation that lets me tweak the spread of the 
numbers. I'm going to make that data, plot it, and then we'll
look at enrichment from both sides.

```{r, slight-squish-effect-gsea, cache=T}
fake_datar_full <- makeFakeData(bias=-0.5,bias_sd=-0.99,
  terms_to_bias=c("A","B","C"))
g(fake_datar_full)

fake_datar <- setNames(object=fake_datar_full$Measurement,
    nm=fake_datar_full$GeneID)

GSEA_result_high <- clusterProfiler::GSEA(
  geneList=sort(fake_datar,decreasing=T),
  nPerm=1e5,
  pvalueCutoff=0.05,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
as.tibble(GSEA_result_high) 

GSEA_result_low <- clusterProfiler::GSEA(
  geneList=sort(-fake_datar,decreasing=T),
  nPerm=1e5,
  pvalueCutoff=0.05,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=term_gene_mapping,
  TERM2NAME=NA)
as.tibble(GSEA_result_low) 
```

So a significant score is that it's biased to one side or the other
on rankings, and the sign of the score is which side it's enriched
on. So you should be able to do one test (lots of permutations) and
that'll test both sides of the enrichment.

Note that `P` is shifted up in enrichment, just by chance. P-value
cutoffs exist for a reason, and are not foolproof. Multiple lines of
evidence are better than one well-tuned experiment, and robustness
of the effect matters more.

# Use with real GO terms

Now we're going to apply it to a real dataset, with real GO terms
(yeast).
You're going to need some mapping of gene to terms, so we're going
to read in 
[the SGD GFF3](https://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff).
Then, from 
[curated data folders](https://downloads.yeastgenome.org/curation/)
we'll look in the `literature` folder, and you can find
a few tab delimited files with genes associated with properties.
You don't have to do just GO terms, but here we'll take a look at
the three GO term files - GO, GO slim[^slim], and GO complexes.
I'm tracking these in this git repo for ease, and I don't expect
them to be updated frequently. Download your own for your own use.

[^slim]: More information on GO slims 
[http://geneontology.org/page/go-slim-and-subset-guide](http://geneontology.org/page/go-slim-and-subset-guide). 
Remember, GO terms aren't holy documents, they're made by people and
subject to choices along the way, remember 
[Stamp's law](https://en.wikipedia.org/wiki/Josiah_Stamp,_1st_Baron_Stamp#Quotes).

Here, I read these in and parse them into 
[tibbles](http://tibble.tidyverse.org/). Sorry that it's so 
convoluted without documentation. If you're keen on learning, you
could step each one through and read about each step and test it
to learn a bit of munge-fu. Or just copy-paste.

```{r,get-tables-input,cache=T,message=F,error=F,warning=F}
terms_per_gene <- read_tsv("data/saccharomyces_cerevisiae.gff",
    comment="#",col_names=F) %>% 
  pull(.,X9) %>% 
  {tibble(X1=str_extract(.,pattern="ID=[^;]+;"),  
    X2=str_extract(.,pattern="Ontology_term=[^;]+;"))} %>%
  filter(!is.na(X2)) %>%
  mutate_all(.,.funs=c(function(x){
      str_remove(x,";")%>%str_remove(.,"ID=")%>%
        str_remove(.,"Ontology_term=")%>%str_remove_all(.,"GO:")
    }))%>%
  mutate(X2=str_split(X2,pattern=","))%>%
  mutate(X2=map(X2,as.integer))%>%
  unnest(X2) %>%
  rename(Systematic=X1,Term=X2) %>%
  mutate(Term=as.character(Term))%>%
  left_join(
    .,
    read_tsv("data/go_terms.tab",
        col_names=c("Term","Description","Ontology","LongDescription"))%>%
      nest(Description,Ontology,LongDescription) %>% 
      mutate(Term=as.character(Term))
    ,
    by="Term") %>%
  mutate(.,data_null=unlist(map(data,is.null))) %>%
  filter(.,!data_null)%>% select(-data_null)%>%
  unnest(data) 

head(terms_per_gene)

go_slim_tibble <- read_tsv("data/go_slim_mapping.tab" ,
    col_names=c("Systematic","Common","SGDID","Ontology","Term","GOID","Type"))%>%
  mutate(GOID=as.integer(str_remove(GOID,"GO:")))

head(go_slim_tibble)

go_complex_tibble <- read_tsv("data/go_protein_complex_slim.tab",
    col_names=F)%>%
  mutate(.,X1=str_remove(X1,"Component: ")) %>%
  mutate(.,X2=str_split(X2,"\\/\\|"))%>%
  mutate(.,X2=sapply(X2,function(x){
      str_replace(x,pattern=".*?\\/[^\\/]+\\/([^\\/]+)\\/.*",replacement="\\1")
    }))%>%
  unnest(X2) %>%
  rename(Systematic=X2,Complex=X1)%>%select(2,1)

head(go_complex_tibble)

# This is a handy table for mapping systematic names to SGD ID
systematic_common  <- go_slim_tibble%>%
  select(.,Systematic,Common)%>%unique(.)%>%
  mutate(.,Common=ifelse(is.na(Common),Systematic,Common))

systematic_common[1000:1010,]
```

Here's the data. It's from the Barseq after FACS after FISH, aka
the Great Red Herring Fishing Expedition of 2015-2017. 

```{r,dataToAnalyze,cache=T,message=F,warning=F}
datar <- read_csv("data/Figure4_Table_BFFfilteredPooledModels.csv",
  comment="#",
  col_names=c("Systematic","Common","PreShift_Mean",
    "PostShift_Mean","Change_Mean","ResidualFromExpected_Mean"))
```

The data has a few columns. Namely, there's the mean signal before
and after a shift. What does this sorta look like? 

## Categorical GO

Well, we could
look at the distribution of this mean after the shift for all
strains, and draw a line at 7.5.

```{r,histogram,cache=T,warning=F}
ggplot(datar)+aes(x=PostShift_Mean)+geom_histogram(bins=50)+
  geom_vline(xintercept=7.5)
```

## Categorical GO

Having done so, we can look for terms enriched in that set. So we're
going to use joins to make our term mapper, and test in a couple of
ways. We're going to try first the whole map, then the GO slim set,
then complexes.

Note that we use `intersect` to only consider the genes that are
in the `datar` dataset.

```{r,real-categorical-enrichment,cache=T}
gene_list <- as.vector(unlist(na.omit(
    datar[datar$PostShift_Mean>7.5,"Systematic"]
  )))

enricher_full <- clusterProfiler::enricher(
  gene=gene_list,
  pvalueCutoff=0.10,
  qvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  universe=intersect(terms_per_gene$Systematic,datar$Systematic),
  TERM2GENE=terms_per_gene %>%select(Description,Systematic)
  )
as.tibble(enricher_full)

enricher_slim <- clusterProfiler::enricher(
  gene=gene_list,
  pvalueCutoff=0.10,
  qvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  universe=intersect(go_slim_tibble$Systematic,datar$Systematic),
  TERM2GENE=go_slim_tibble %>%select(Term,Systematic)
  )
as.tibble(enricher_slim)

enricher_comp <- clusterProfiler::enricher(
  gene=gene_list,
  pvalueCutoff=0.10,
  qvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  universe=intersect(go_complex_tibble$Systematic,datar$Systematic),
  TERM2GENE=go_complex_tibble %>%select(Complex,Systematic)
  )
as.tibble(enricher_comp)
```

We can see that the slim categories are very general. 

We can also look using GSEA. This takes a long time, so I'm
breaking these into independently cached chunks.

```{r, real-gsea-enrichment-1, cache=T}
genes_ordered <- sort(
  na.omit(setNames(datar$PostShift_Mean, nm=datar$Systematic)),
  decreasing=T)

nperms <- 1e5
```

```{r, real-gsea-enrichment-2, cache=T}
GSEA_full <- clusterProfiler::GSEA(
  geneList=genes_ordered,
  nPerm=nperms,
  pvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=terms_per_gene %>%select(Description,Systematic)
  )
as.tibble(GSEA_full)
```

```{r, real-gsea-enrichment-3, cache=T}
GSEA_slim <- clusterProfiler::GSEA(
  geneList=genes_ordered,
  nPerm=nperms,
  pvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=go_slim_tibble %>%select(Term,Systematic)
  )
as.tibble(GSEA_slim)
```

```{r, real-gsea-enrichment-4, cache=T}
GSEA_comp <- clusterProfiler::GSEA(
  geneList=genes_ordered,
  nPerm=nperms,
  pvalueCutoff=0.10,
  pAdjustMethod="fdr",
  minGSSize=1,maxGSSize=500,
  TERM2GENE=go_complex_tibble %>%select(Complex,Systematic)
  )
as.tibble(GSEA_comp)
```

## Some plots

I think the table (or tibble) is the most important thing, but if you
want to visualize it then the author had made a few tools for that.
I found that `barplot` and `emapplot` were not findable on my system.
The below examples are just lifted from the BioC vingette, so more
details there.

```{r,someplots,cache=T}
DOSE::dotplot(GSEA_full)
DOSE::cnetplot(GSEA_full,categorySize="qvalue",foldChange=genes_ordered)
```

One of the handiest might be to go ahead and take the ID of the gene
set, and use the `gseaPlot` to look at how that enrichment got
calculated:

```{r,moreplots,cache=T}
as.tibble(GSEA_full)
DOSE::gseaplot(GSEA_full,geneSetID="Lsm1-7-Pat1 complex")
DOSE::gseaplot(GSEA_full,geneSetID="negative regulation of gluconeogenesis")
```

Anyways, there's a lot more on the vingette
[here](https://www.bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#visualization-1).

# Further reading

This author has made some 
[pretty nifty tools](https://guangchuangyu.github.io/software/).
Check out the documentation, now that you're started on using it. 

You may have trouble in parsing a large list of terms for finding 
what's the interesting patterns. If you save the `enricherResult`
object that's returned from the analyses, you should be able to
explore it using the [`GOSemSim`](https://guangchuangyu.github.io/software/GOSemSim/)
package, for collapsing terms based on semantic similarity.
I've not used that (just manually interpreted), so if someone can
figure that out it would be helpful to add your notes to this 
document! And edit!

```{r}
sessionInfo()
```
